### PR TITLE
feat: separate DepositTx and set suggested fee recipient

### PR DESCRIPTION
**Pull request by Pangssu**
To reduce code diffs with optimism, separate DepositTx into a OP version and Kroma version, and set suggestedFeeRecipient for payload attributes

This PR cannot use the geth that includes predeploy contract address changes introduced in https://github.com/kroma-network/go-ethereum/pull/125. Therefore, it references commit `fd7e2919` from the https://github.com/kroma-network/go-ethereum/commits/pang/separate-deposit-tx branch. The devnet L2 docker image is also built based on this commit.

### CODE REVIEW

File: op-node/p2p/sync.go [hunk 1]
```diff
@@ -668,6 +670,20 @@ func (s *SyncClient) readExecutionPayload(data []byte, expectedTime uint64) (*et
...
+				var err error
+				res.Transactions[i], err = derive.ToKromaDepositBytes(otx)
+				if err != nil {
+					return nil, fmt.Errorf("failed to convert deposit tx: %w", err)
+				}
+			}
+		}
+	}
+	// [Kroma: END]
+
```
Comment (kangsorang) [inline op-node/p2p/sync.go:686]: Since we already have ecotone applied, shouldn't we consider the case for version 1 ? I don't think there are any p2p nodes requesting version 0.

File: op-service/eth/types.go [hunk 2]
```diff
@@ -261,6 +261,15 @@ func BlockAsPayload(bl *types.Block, canyonForkTime *uint64) (*ExecutionPayload,
 			return nil, fmt.Errorf("tx %d failed to marshal: %w", i, err)
 		}
 		opaqueTxs[i] = otx
+		// [Kroma: START] Use KromaDepositTx instead DepositTx
```
Comment (kangsorang) [inline op-service/eth/types.go:264]: Suddenly I had this thought. 
Couldn't we just apply IsKromaDepositTx to encodeTyped paired with the decodeTyped function and not have to put this code in?

File: packages/contracts/scripts/go-ffi/differential-testing.go
[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/scripts/go-ffi/differential-testing.go:1]: IMO, add isKromaMPT arg to encodeDepositTransaction and hashDepositTransaction, and then handle both cases here as getProveWithdrawalTransactionInputs does

File: packages/contracts/scripts/go-ffi/differential-testing.go [hunk 3]
```diff
@@ -176,9 +176,10 @@ func DiffTestUtils() {
 
 		// Create deposit transaction
 		depositTx := makeDepositTx(from, to, value, mint, gasLimit, false, data, l1BlockHash, logIndex)
+		kromaDepTx := toKromaDepositTx(depositTx)
 
 		// RLP encode deposit transaction
-		encoded, err := types.NewTx(&depositTx).MarshalBinary()
+		encoded, err := types.NewTx(&kromaDepTx).MarshalBinary()
 		checkErr(err, "Error encoding deposit transaction")
```
Comment (seolaoh) [inline packages/contracts/scripts/go-ffi/differential-testing.go:183]: Additional function `toKromaDepositTx` is unnecessary
```suggestion

		// RLP encode deposit transaction
		encoded, err := types.NewTx(&depositTx).MarshalBinary()
		checkErr(err, "Error encoding deposit transaction")

		// [Kroma: START]
		encoded, err = derive.ToKromaDepositBytes(encoded)
		checkErr(err, "Error converting deposit transaction to KromaDepositTx")
		// [Kroma: END]
```

File: op-node/rollup/derive/attributes.go [hunk 4]
```diff
@@ -130,14 +149,34 @@ func (ba *FetchingAttributesBuilder) PreparePayloadAttributes(ctx context.Contex
 		}
 	}
 
+	suggestedFeeRecipient := common.Address{}
+	if ba.rollupCfg.IsKromaMPT(nextL2Time) {
+		suggestedFeeRecipient = predeploys.ProtocolVaultAddr
+	}
```
Comment (seolaoh) [inline op-node/rollup/derive/attributes.go:155]: Wrap here with Kroma

File: op-node/rollup/derive/attributes_queue_test.go [hunk 5]
```diff
@@ -11,6 +11,7 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/log"
 
+	"github.com/ethereum-optimism/optimism/kroma-bindings/predeploys"
```
Comment (seolaoh) [inline op-node/rollup/derive/attributes_queue_test.go:14]:
```suggestion
	"github.com/kroma-network/kroma/kroma-bindings/predeploys"
```

File: op-node/rollup/derive/attributes_queue_test.go [hunk 6]
```diff
@@ -64,26 +60,49 @@ func TestAttributesQueue(t *testing.T) {
...
+		require.NoError(t, err)
+		require.Equal(t, attrs, *actual)
 	}
-	attrBuilder := NewFetchingAttributesBuilder(cfg, l1Fetcher, l2Fetcher)
 
-	aq := NewAttributesQueue(testlog.Logger(t, log.LevelError), cfg, attrBuilder, nil)
+	t.Run("before kroma mpt time", func(st *testing.T) {
+		rollupCfg := rollup.Config{}
+		l1InfoTx, err := L1InfoDepositBytes(&rollupCfg, expectedL1Cfg, safeHead.SequenceNumber+1, l1Info, 0)
+		require.NoError(t, err)
```
Comment (seolaoh) [inline op-node/rollup/derive/attributes_queue_test.go:91]:
```suggestion
		require.NoError(st, err)
```

File: op-node/rollup/derive/attributes_queue_test.go [hunk 7]
```diff
@@ -64,26 +60,49 @@ func TestAttributesQueue(t *testing.T) {
...
-	require.NoError(t, err)
-	require.Equal(t, attrs, *actual)
+	t.Run("after kroma mpt time", func(st *testing.T) {
+		zero := uint64(0)
+		cfg.KromaMPTTime = &zero
+		rollupCfg := rollup.Config{
+			KromaMPTTime: &zero,
+		}
+		l1InfoTx, err := L1InfoDepositBytes(&rollupCfg, expectedL1Cfg, safeHead.SequenceNumber+1, l1Info, 0)
+		require.NoError(t, err)
```
Comment (seolaoh) [inline op-node/rollup/derive/attributes_queue_test.go:105]:
```suggestion
		require.NoError(st, err)
```

File: op-service/eth/types.go (hunk 2 above)
Comment (Pangssu) [reply op-service/eth/types.go:264]: This code is necessary because DepositTx should only be applied after MPT time.

File: op-node/rollup/derive/ecotone_upgrade_transactions.go [hunk 8]
```diff
@@ -53,8 +53,8 @@ func EcotoneNetworkUpgradeTransactions() ([]hexutil.Bytes, error) {
 		Data:       l1BlockDeploymentBytecode,
 		// [Kroma: START]
 		Gas: 500_000,
-		// IsSystemTransaction: false,
 		// [Kroma: END]
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply op-node/rollup/derive/ecotone_upgrade_transactions.go:56]: Impossible. When syncing from past blocks, the gas must remain 500,000.

File: packages/contracts/contracts/test/Encoding.t.sol [hunk 9]
```diff
@@ -56,6 +56,7 @@ contract Encoding_Test is CommonTest {
         assertEq(encoding, _encoding);
     }
 
+    // [Kroma: START]
```
Comment (sm-stack) [inline packages/contracts/contracts/test/Encoding.t.sol:59]: nit: wouldn't it be better to lower this to `// assert DepositTx part` (L#81)?

**Review by sm-stack — APPROVED** [review]

File: packages/contracts/contracts/test/Encoding.t.sol (hunk 9 above)
Comment (Pangssu) [reply packages/contracts/contracts/test/Encoding.t.sol:59]: I marked it at the function level, like `Hashing.t.sol`.

File: op-node/rollup/derive/l1_block_info.go [hunk 10]
```diff
@@ -352,10 +352,5 @@ func L1InfoDepositBytes(rollupCfg *rollup.Config, sysCfg eth.SystemConfig, seqNu
 	if err != nil {
 		return nil, fmt.Errorf("failed to encode L1 info tx: %w", err)
 	}
-	// [Kroma: START] Use KromaDepositTx instead of DepositTx before MPT time
-	if !rollupCfg.IsKromaMPT(l2BlockTime) {
-		opaqueL1Tx, err = ToKromaDepositBytes(opaqueL1Tx)
-	}
-	// [Kroma: END]
```
Comment (sm-stack) [inline op-node/rollup/derive/l1_block_info.go]: Why is this part removed?

File: op-e2e/op_geth_test.go [hunk 11]
```diff
@@ -473,6 +459,10 @@ func TestRegolith(t *testing.T) {
 			// L1Info tx should report actual gas used, not 0 or the tx gas limit
 			infoTx, err := opGeth.L2Client.TransactionInBlock(ctx, envelope.ExecutionPayload.BlockHash, 0)
 			require.NoError(t, err)
+			// [Kroma: START] Use KromaDepositTx instead of DepositTx
+			infoTx, err = infoTx.ToKromaDepositTx()
+			require.NoError(t, err)
+			// [Kroma: END]
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply op-e2e/op_geth_test.go:489]: applied all DepositTxs

File: op-node/rollup/derive/attributes_test.go [hunk 12]
```diff
@@ -113,6 +113,10 @@ func TestPreparePayloadAttributes(t *testing.T) {
 		l1Info.InfoNum = l2Parent.L1Origin.Number + 1
 		epoch := l1Info.ID()
 		l1InfoTx, err := L1InfoDepositBytes(cfg, testSysCfg, 0, l1Info, 0)
+		// [Kroma: START] Use KromaDepositTx instead of DepositTx
+		l1InfoTx, err = ToKromaDepositBytes(l1InfoTx)
+		require.NoError(t, err)
+		// [Kroma: END]
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply op-node/rollup/derive/attributes_test.go:119]: added too

**Review by kangsorang — APPROVED** [review]
LGTM

File: op-e2e/op_geth.go [hunk 13]
```diff
@@ -220,6 +220,20 @@ func (d *OpGeth) CreatePayloadAttributes(txs ...*types.Transaction) (*eth.Payloa
 		txBytes = append(txBytes, bin)
 	}
 
+	// [Kroma: START] Use KromaDepositTx instead of DepositTx
+	if d.L2ChainConfig.IsPreKromaMPT(uint64(timestamp)) {
+		for i := range txBytes {
+			if txBytes[i][0] == types.DepositTxType {
+				bin, err := derive.ToKromaDepositBytes(txBytes[i])
+				if err != nil {
+					return nil, fmt.Errorf("failed to convert DepositTx to KromaDepositTx")
```
Comment (seolaoh) [inline op-e2e/op_geth.go:229]:
```suggestion
					return nil, fmt.Errorf("failed to convert DepositTx to KromaDepositTx: %w", err)
```

File: op-e2e/op_geth_test.go [hunk 14]
```diff
@@ -1061,3 +1095,118 @@ func TestEcotone(t *testing.T) {
...
+	tests := []struct {
+		name        string
+		kromaMPTTime *hexutil.Uint64
+	}{
+		{name: "KromaMPTNotScheduled"},
+		{name: "KromaMPTNotYetActive", kromaMPTTime: &futureTimestamp},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(fmt.Sprintf("KromaDepositTx%s", test.name), func(t *testing.T) {
```
Comment (seolaoh) [inline op-e2e/op_geth_test.go:1111]:
```suggestion
		t.Run(fmt.Sprintf("KromaDepositTx_%s", test.name), func(t *testing.T) {
```

File: op-e2e/op_geth_test.go [hunk 15]
```diff
@@ -1061,3 +1095,118 @@ func TestEcotone(t *testing.T) {
...
+func TestKromaMPT(t *testing.T) {
+	tests := []struct {
+		name            string
+		kromaMPTTime     hexutil.Uint64
+	}{
+		{name: "ActivateAtGenesis", kromaMPTTime: 0},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(fmt.Sprintf("ReviveDepositTx%s", test.name), func(t *testing.T) {
```
Comment (seolaoh) [inline op-e2e/op_geth_test.go:1169]:
```suggestion
		t.Run(fmt.Sprintf("ReviveDepositTx_%s", test.name), func(t *testing.T) {
```

File: op-node/p2p/sync.go [hunk 16]
```diff
@@ -670,20 +668,6 @@ func (s *SyncClient) readExecutionPayload(data []byte, expectedTime uint64) (*et
...
-				var err error
-				res.Transactions[i], err = derive.ToKromaDepositBytes(otx)
-				if err != nil {
-					return nil, fmt.Errorf("failed to convert deposit tx to KromaDepositTx: %w", err)
-				}
-			}
-		}
-	}
-	// [Kroma: END]
-
```
Comment (sm-stack) [inline op-node/p2p/sync.go]: Curious, why is this part removed?

File: op-node/rollup/derive/engine_queue_test.go [hunk 17]
```diff
@@ -968,6 +968,12 @@ func TestBlockBuildingRace(t *testing.T) {
 		InfoBloom:            types.Bloom{},
 		InfoExtra:            []byte{},
 	}, 0)
+	// [Kroma: START] Use KromaDepositTx instead of DepositTx
+	if cfg.KromaMPTTime == nil || !cfg.IsKromaMPT(refA1.Time + cfg.BlockTime) {
```
Comment (seolaoh) [inline op-node/rollup/derive/engine_queue_test.go:972]:
```suggestion
	if cfg.KromaMPTTime == nil || !cfg.IsKromaMPT(refA1.Time) {
```

**Review by seolaoh — APPROVED** [review]

File: op-node/p2p/sync.go (hunk 16 above)
Comment (Pangssu) [reply op-node/p2p/sync.go]: This part handles unsafe payload received via p2p. 
Since the payload delivered by peers will already include deposit txs converted to KromaDepositTx, this conversion in this part has been removed.

**Review by sm-stack — APPROVED** [review]
